### PR TITLE
EZP-25128: Automatically navigate to 1st item of a zone

### PR DESCRIPTION
**Pull request by yannickroger**
Link: https://jira.ez.no/browse/EZP-25128

## Description
Before this patch. When clicking on a top menu item (like "Content") sub items of this menu would be displayed. With this patch, on top of the existing behavior, the first item is automatically opened.

Also, I fixed a few typos I ran into in separate commits.

## Tests
Manual and automated

### CODE REVIEW

**Comment by yannickroger** [comment]
@dpobel @mhyndle ready for review.

File: Resources/public/js/views/services/ez-navigationhubviewservice.js [hunk 1]
```diff
@@ -53,6 +54,28 @@ YUI.add('ez-navigationhubviewservice', function (Y) {
             });
         },
 
+        /**
+         * activeNavigationChange event handler, it redirects to the first item of the clicked zone
+         *
+         * @method _activeNavigationChange
+         */
```
Comment (mhyndle) [inline Resources/public/js/views/services/ez-navigationhubviewservice.js]: missing param docs

Comment (yannickroger) [inline Resources/public/js/views/services/ez-navigationhubviewservice.js]: Damn, I often ask people for those params, and I forget them :)

File: Tests/js/views/services/assets/ez-navigationhubviewservice-tests.js [hunk 2]
```diff
@@ -895,6 +896,79 @@ YUI.add('ez-navigationhubviewservice-tests', function (Y) {
...
+                        params: this.routeParams,
+                    }
+                }
+            };
+
+            this.appMock = new Mock();
+
+            Mock.expect(this.appMock, {
+                method: 'navigateTo',
+                args: [Mock.Value.String, Mock.Value.Any],
```
Comment (mhyndle) [inline Tests/js/views/services/assets/ez-navigationhubviewservice-tests.js]: You can simplify it a bit with something like:
`args: [that.routeName, that.routeParams]`
so no need of defining `run` then

File: Tests/js/views/services/assets/ez-navigationhubviewservice-tests.js [hunk 3]
```diff
@@ -895,6 +896,79 @@ YUI.add('ez-navigationhubviewservice-tests', function (Y) {
...
+
+            this.view = new Y.View({
+                'bubbleTargets': this.service,
+                'activeNavigation': ''
+            });
+        },
+
+        tearDown: function () {
+            this.service.destroy();
+            delete this.service;
```
Comment (mhyndle) [inline Tests/js/views/services/assets/ez-navigationhubviewservice-tests.js]: Destroy and delete `this.view` also?

File: Tests/js/views/services/assets/ez-navigationhubviewservice-tests.js (hunk 2 above)
Comment (mhyndle) [inline Tests/js/views/services/assets/ez-navigationhubviewservice-tests.js]: Edit: just `Mock.verify(this.appMock)`  to verify if `nvigateTo` was called.

File: Resources/public/js/views/services/ez-navigationhubviewservice.js [hunk 4]
```diff
@@ -23,6 +23,7 @@ YUI.add('ez-navigationhubviewservice', function (Y) {
         initializer: function () {
             this.on('*:logOut', this._logOut);
             this.after('*:navigateTo', this._navigateTo);
+            this.after('*:activeNavigationChange', this._activeNavigationChange);
```
Comment (dpobel) [inline Resources/public/js/views/services/ez-navigationhubviewservice.js]: `_activeNavigationChange` is a very bad name as it gives absolutely no clue of it is doing. Good method names usually start with a verb and tell what the method does, not why/when it is called. So here for instance, the name could be `_navigateToFirstItemRoute`. That way, you can almost read that line as an English sentence.

**Comment by yannickroger** [comment]
@mhyndle 's feedback taken into account.

File: Resources/public/js/views/services/ez-navigationhubviewservice.js (hunk 4 above)
Comment (yannickroger) [inline Resources/public/js/views/services/ez-navigationhubviewservice.js]: That used to be its name… then I followed the examples of the ones above, named after the even :( I'll fix it.

File: Resources/public/js/views/services/ez-navigationhubviewservice.js [hunk 5]
```diff
@@ -53,6 +54,31 @@ YUI.add('ez-navigationhubviewservice', function (Y) {
...
+         */
+        _activeNavigationChange: function (e) {
+            var zone = this.get(e.newVal + 'NavigationItems'),
+                firstItem,
+                route;
+
+            if (zone && zone.length > 0) {
+                firstItem = zone[0];
+
+                if (firstItem.config) {
```
Comment (dpobel) [inline Resources/public/js/views/services/ez-navigationhubviewservice.js]: we already have the same kind of test to detect whether it's a config or if we already have a true navigation item view and this is done by checking the `get` method in https://github.com/ezsystems/PlatformUIBundle/blob/master/Resources/public/js/views/services/ez-navigationhubviewservice.js#L343 and https://github.com/ezsystems/PlatformUIBundle/blob/master/Resources/public/js/views/services/ez-navigationhubviewservice.js#L319 so I guess it's time to introduce a small private method to do that.

File: Tests/js/views/services/assets/ez-navigationhubviewservice-tests.js [hunk 6]
```diff
@@ -895,6 +896,83 @@ YUI.add('ez-navigationhubviewservice-tests', function (Y) {
...
+            this.view.destroy();
+            delete this.view;
+        },
+
+        "Should not do anything when no item": function () {
+            this.view.set('activeNavigation', this.zoneName);
```
Comment (dpobel) [inline Tests/js/views/services/assets/ez-navigationhubviewservice-tests.js]: while the code appears to be covered by that, this is incomplete. Before this line, you should (re)configure `appMock` to not expect any call to `navigateTo` and then call `Mock.verify` at the end.

**Comment by yannickroger** [comment]
@dpobel 's feedback taken into account

File: Resources/public/js/views/services/ez-navigationhubviewservice.js [hunk 7]
```diff
@@ -53,6 +54,44 @@ YUI.add('ez-navigationhubviewservice', function (Y) {
...
         },
 
+
+        /**
+         * Returns true if a provided item is a NavigationItemView
+         *
+         * @method _isNavigationViewItem
+         * @private
+         * @param {Object} item
+         * @return true if item is a NavigationItemView, false if item is a NavigationItemParameterView
```
Comment (dpobel) [inline Resources/public/js/views/services/ez-navigationhubviewservice.js]: `@return {Boolean}`

if more than enough :)

File: Resources/public/js/views/services/ez-navigationhubviewservice.js [hunk 8]
```diff
@@ -53,6 +54,44 @@ YUI.add('ez-navigationhubviewservice', function (Y) {
...
+         */
+        _isNavigationViewItem: function (item) {
+            return item.get;
```
Comment (dpobel) [inline Resources/public/js/views/services/ez-navigationhubviewservice.js]: Nitpick: if you want it to return a boolean it should actually be `return !!item.get;`

File: Resources/public/js/views/services/ez-navigationhubviewservice.js [hunk 9]
```diff
@@ -53,6 +54,44 @@ YUI.add('ez-navigationhubviewservice', function (Y) {
...
+        },
+
+        /**
+         * Redirects to the first item of the clicked zone
```
Comment (dpobel) [inline Resources/public/js/views/services/ez-navigationhubviewservice.js]: nitpick: s/clicked/active/

**Comment by dpobel** [comment]
besides the 3 nitpicks, +1

**Comment by yannickroger** [comment]
" @dpobel 's feedback strike back " taken into account, ping @mhyndle

**Comment by dpobel** [comment]
more "@dpobel awakens" today ;-)

**Comment by yannickroger** [comment]
I was afraid of having a " Return of the @dpobel "

**Comment by mhyndle** [comment]
just one thing about my comment: https://github.com/ezsystems/PlatformUIBundle/pull/480/files#r47629741
`run` is not doing anything else than making assertions which may be bit simplified with what I have proposed in the comment (or it's not valid?)
+1, besides above

**Comment by yannickroger** [comment]
Sorry @mhyndle I missed it. It's taken care of now.

**Comment by yannickroger** [comment]
Can this be merged with the broken BDD @bdunogier @andrerom @pcardiga or can you can (help me? ) fix it ?
I am still a bit lost with the new CI BDD system.

**Comment by miguelcleverti** [comment]
@yannickroger there seems to be some problem with the roles behat tests, I am looking into it. However it does not seem related to this PR.

**Comment by bdunogier** [comment]
I have executed behat (`./bin/behat --profile=platformui vendor/ezsystems/platform-ui-bundle/Features/Users/users.feature --tags='common'`) on my local instance, and it seeems that something is wrong.

Without the patch, ``Create a valid user` passes.
With it, it fails with a timeout. I have added a bit more data, catching the exception thrown by `spin()` from `fillFieldWithValue`: `Timeout while waiting for loading element '.is-app-loading, .is-app-transitioning, .yui3-app-transitioning'. (Exception)`.

**Comment by miguelcleverti** [comment]
It seems that what is happening is that behat is interacting with the page before the automatic navigation to the 1st item of the zone.
This causes failures because behat expects something, but the page changed meanwhile.
The tests need to be updated to take this into account. @bdunogier @yannickroger

**Comment by dpobel** [comment]
@miguelcleverti your patch (https://github.com/ezsystems/PlatformUIBundle/commit/941d718f2eb79f286394e7f8d338b48841c1ffb6) does not seem enough. Can you have another look please ? (and please do not open a new pull request, work directly on the ezp-25128-content_menu_link branch)

**Comment by miguelcleverti** [comment]
A problem seems to have escaped me before! When accessing directly to a page URL instead of interacting with the UI causes the load of the first item. Ex:
Using the Role url `http://ezp6.dev/ez#/admin/pjax%2Frole` instead navigating through `Administration dashboard` > `Roles` in the UI.
IMO I don't think this should happen only when navigating to a zone should it load the first item.

**Comment by dpobel** [comment]
@miguelcleverti you're perfectly right :) After all it's useful to have automated tests :D
@yannickroger can you have a look ? (I know it's been a long time...)

**Comment by yannickroger** [comment]
This kind of long time is relative when you are used to troubleshooting legacy :)
I will have a look soon.

**Comment by yannickroger** [comment]
Issue fixed in a dedicated commit. ping @dpobel @StephaneDiot

File: .ez.yml [hunk 10]
```diff
@@ -4,7 +4,7 @@ v: 0
 meta:
   repo: "https://github.com/ezsystems/ezplatform.git"
   branch: "master"
-  self_alias: "1.0.x-dev"
+  self_alias: "1.2.x-dev"
```
Comment (bdunogier) [inline .ez.yml]: Against master, it would actually be `1.3.x-dev` (same as `master`).

**Comment by StephaneDiot** [comment]
+1

File: Resources/public/js/views/ez-navigationhubview.js [hunk 11]
```diff
@@ -130,7 +130,10 @@ YUI.add('ez-navigationhubview', function (Y) {
                     activeZone = key;
                 }
             }, this);
-            this.set('activeNavigation', activeZone);
+
+            if (this.get('activeNavigation') !== null) {
```
Comment (dpobel) [inline Resources/public/js/views/ez-navigationhubview.js]: I think this is the wrong place for this condition. Here the `activeNavigation` attribute should always be set. But in the `activeNavigationChange` handler in the NavigationHubViewService, you should check the previous `activeNavigation` value to trigger the first route navigation only if there was a `activeNavigation` before, meaning the user has chosen a zone.

**Comment by yannickroger** [comment]
@dpobel 's feedback taken into account

**Comment by dpobel** [comment]
+1

**Pull request merged**